Sources/IBMSwiftSDKCore/Authentication/IAMAuthenticator.swift

```diff
@@ -70,28 +73,27 @@ class IAMTokenSource: TokenSource {
     }
 
     func getToken(completionHandler: @escaping (String?, RestError?) -> Void) {
-        // request a new access token if the current token is expired
-        guard let refreshDate = refreshDate, refreshDate.timeIntervalSinceNow > 0 else {
-            requestToken {
-                token, error in
-
-                guard let token = token, error == nil else {
-                    completionHandler(nil, error)
-                    return
-                }
-                self.token = token.accessToken
-                // Compute refreshDate for token
-                let expirationDate = Date(timeIntervalSince1970: Double(token.expiration))
-                // We want a little buffer to make sure we refresh proactively
-                let buffer = expirationDate.timeIntervalSinceNow * -0.2
-                self.refreshDate = expirationDate.addingTimeInterval(buffer)
-                completionHandler(self.token, nil)
+        // If we have a token, pass it to the completion handler.
+        if let token = self.token {
+            completionHandler(token, nil)
+            if let refreshDate = refreshDate, refreshDate.timeIntervalSinceNow < 0 {
+                refreshQueue.async{ self.refreshToken() }
             }
+        } else {
+            requestToken(completionHandler: completionHandler)
```

Comment:
Isn't there a problem with this `requestToken` being async causing parallel token requests before the first token is acquired?
That is, if multiple threads call `getToken` before `self.token` is set then they will each end up calling `requestToken`. Maybe the `requestToken` needs to go on the queue as well and recheck if the token is unset before requesting it - although having said that `background` qos is might be insufficient for the first request.

Reply:
Probably right.  I suppose we can just create a new queue for this.